lib/postgres.go

```diff
@@ -0,0 +1,1414 @@
+		}
+	}
+
+	if len(transactions) > 0 {
+		if _, err := tx.Model(&transactions).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(transactionOutputs) > 0 {
+		if _, err := tx.Model(&transactionOutputs).Returning("NULL").OnConflict("(output_hash, output_index) DO UPDATE").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(transactionInputs) > 0 {
+		if _, err := tx.Model(&transactionInputs).WherePK().Column("input_hash", "input_index", "spent").Update(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataBlockRewards) > 0 {
+		if _, err := tx.Model(&metadataBlockRewards).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataBitcoinExchanges) > 0 {
+		if _, err := tx.Model(&metadataBitcoinExchanges).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataPrivateMessages) > 0 {
+		if _, err := tx.Model(&metadataPrivateMessages).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataSubmitPosts) > 0 {
+		if _, err := tx.Model(&metadataSubmitPosts).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataUpdateProfiles) > 0 {
+		if _, err := tx.Model(&metadataUpdateProfiles).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataExchangeRates) > 0 {
+		if _, err := tx.Model(&metadataExchangeRates).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataFollows) > 0 {
+		if _, err := tx.Model(&metadataFollows).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataLikes) > 0 {
+		if _, err := tx.Model(&metadataLikes).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataCreatorCoins) > 0 {
+		if _, err := tx.Model(&metadataCreatorCoins).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataCreatorCoinTransfers) > 0 {
+		if _, err := tx.Model(&metadataCreatorCoinTransfers).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	if len(metadataSwapIdentities) > 0 {
+		if _, err := tx.Model(&metadataSwapIdentities).Returning("NULL").Insert(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (postgres *Postgres) UpsertBlockAndTransactions(blockNode *BlockNode, bitcloutBlock *MsgBitCloutBlock) error {
+	return postgres.db.RunInTransaction(postgres.db.Context(), func(tx *pg.Tx) error {
+		err := postgres.UpsertBlockTx(tx, blockNode)
+		if err != nil {
+			return err
+		}
+
+		blockHash := blockNode.Hash
+		err = postgres.UpsertChainTx(tx, "main", blockHash)
+		if err != nil {
+			return err
+		}
+
+		err = postgres.InsertTransactionsTx(tx, bitcloutBlock.Txns, blockHash)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (postgres *Postgres) GetUtxoEntryForUtxoKey(utxoKey *UtxoKey) *UtxoEntry {
+	utxo := &PGTransactionOutput{
+		OutputHash:  &utxoKey.TxID,
+		OutputIndex: utxoKey.Index,
+		Spent:       false,
```

Comment:
Curious: Are we keeping spent UTXOs around? Makes sense if we are, but this would be different than what we were doing before.

Reply:
I was going to combine Transaction Outputs and UTXOs but this will actually make warp sync more difficult. I'm going to separate these out.